File: client/lib/models/community_review.dart
/// Community book review model.
class CommunityReview {
  final String reviewId;
  final String userId;
  final String catalogBookId;
  final String? authorDisplayName;
  final String? authorUsername;
  final String? authorAvatarUrl;
  final String? title;
  final String body;
  final bool containsSpoilers;
  final String visibility;
  final int likeCount;
  final int helpfulCount;
  final DateTime? createdAt;

  const CommunityReview({
    required this.reviewId,
    required this.userId,
    required this.catalogBookId,
    this.authorDisplayName,
    this.authorUsername,
    this.authorAvatarUrl,
    this.title,
    required this.body,
    this.containsSpoilers = false,
    this.visibility = 'public',
    this.likeCount = 0,
    this.helpfulCount = 0,
    this.createdAt,
  });

  factory CommunityReview.fromJson(Map<String, dynamic> json) =>
      CommunityReview(
        reviewId: json['review_id'] as String,
        userId: json['user_id'] as String,
        catalogBookId: json['catalog_book_id'] as String,
        authorDisplayName: json['author_display_name'] as String?,
        authorUsername: json['author_username'] as String?,
        authorAvatarUrl: json['author_avatar_url'] as String?,
        title: json['title'] as String?,
        body: json['body'] as String,
        containsSpoilers: json['contains_spoilers'] as bool? ?? false,
        visibility: json['visibility'] as String? ?? 'public',
        likeCount: json['like_count'] as int? ?? 0,
        helpfulCount: json['helpful_count'] as int? ?? 0,
        createdAt: json['created_at'] != null
            ? DateTime.parse(json['created_at'] as String)
            : null,
      );

  Map<String, dynamic> toJson() => {
    'review_id': reviewId,
    'user_id': userId,
    'catalog_book_id': catalogBookId,
    'title': title,
    'body': body,
    'contains_spoilers': containsSpoilers,
    'visibility': visibility,
  };
}

File: server/src/papyrus/schemas/community_review.py
"""Review schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from papyrus.schemas.common import BaseSchema, Pagination


class ReviewResponse(BaseSchema):
    review_id: UUID
    user_id: UUID
    catalog_book_id: UUID
    author_display_name: str | None = None
    author_username: str | None = None
    author_avatar_url: str | None = None
    title: str | None = None
    body: str
    contains_spoilers: bool = False
    visibility: str = "public"
    like_count: int = 0
    helpful_count: int = 0
    created_at: datetime | None = None
    updated_at: datetime | None = None


class ReviewList(BaseModel):
    reviews: list[ReviewResponse]
    pagination: Pagination


class CreateReviewRequest(BaseModel):
    catalog_book_id: UUID
    title: str | None = Field(None, max_length=255)
    body: str = Field(..., min_length=10, max_length=10000)
    contains_spoilers: bool = False
    visibility: str = Field("public", pattern=r"^(public|friends|private)$")


class UpdateReviewRequest(BaseModel):
    title: str | None = Field(None, max_length=255)
    body: str | None = Field(None, min_length=10, max_length=10000)
    contains_spoilers: bool | None = None
    visibility: str | None = Field(None, pattern=r"^(public|friends|private)$")


class ReviewReactionRequest(BaseModel):
    reaction_type: str = Field(..., pattern=r"^(like|helpful)$")
